### PR TITLE
Revert the input pipe in the DuplexStreamPipeAdapter

### PR DESCRIPTION
**Pull request by jkotalik**
The other half of https://github.com/aspnet/AspNetCore/pull/11601

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 1]
```diff
@@ -89,6 +109,40 @@ public override ValueTask DisposeAsync()
...
+                    if (flushResult.IsCompleted)
+                    {
+                        // flushResult should not be canceled.
+                        break;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log?.LogCritical(0, ex, $"{GetType().Name}.{nameof(ReadInputAsync)}");
```
Comment (benaadams) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: Does this ever happen, should it have a `LoggerMessage.Define`? Either in [`KestrelTrace`](https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs) or [`LoggerExtensions`](https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Core/src/Internal/LoggerExtensions.cs)?

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: We have the same pattern in WriteOutputAsync. I don't think we expect this to ever happen. https://github.com/aspnet/AspNetCore/pull/12204/files#diff-23de5f6c27fec1507959dd5e19dfe3f0R186

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 2]
```diff
@@ -50,43 +64,106 @@ internal class DuplexPipeStreamAdapter<TStream> : DuplexPipeStream, IDuplexPipe
...
-                return default;
+                return;
             }
 
-            return new ValueTask(_outputTask);
+            await _outputTask;
+
+            CancelPendingRead();
+
+            await _inputTask;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: Could this nullref if the Input or Output property was never accessed and RunAsync() was never called?

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 3]
```diff
@@ -50,43 +64,106 @@ internal class DuplexPipeStreamAdapter<TStream> : DuplexPipeStream, IDuplexPipe
...
+                    var bytesRead = await Stream.ReadAsync(outputBuffer);
+                    _input.Writer.Advance(bytesRead);
+
+                    if (bytesRead == 0)
+                    {
+                        // FIN
+                        break;
+                    }
+
+                    var result = await _input.Writer.FlushAsync();
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: Is it possible that you could be stuck awaiting this call to flush indefinitely when DisposeAsync is called? Do we need to call CancelPendingFlush in the _input.Writer?

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs (hunk 2 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: Not sure but I'll handle it no matter what.

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs (hunk 3 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: Nvm you completed the _input.Reader which will unblock this call to FlushAsync()

**Review by halter73 — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

**Comment by davidfowl** [comment]
😢

**Pull request merged**